lib/spack/spack/solver/asp.py

```diff
@@ -2046,7 +2048,10 @@ def define_ad_hoc_versions_from_specs(
             # best possible, so they're guaranteed to be used preferentially.
             version = s.versions.concrete
 
-            if version is None or any(v == version for v in self.possible_versions[s.name]):
+            if version is None or (
+                not isinstance(version, vn.GitVersion)
+                and any((v == version) for v in self.possible_versions[s.name])
```

Comment:
Isn't the first constraint redundant since the declared versions can't be GitVersion? Or has another method already stored `GitVersion` constraints in the dict?

Reply:
Prior installs are in the dictionary as well.